File: benchmarks/fixtures/webapp_swift/Sources/webapp/middleware/AuthMiddleware.swift
import Foundation

/// Rejects requests without a valid token before they reach a handler.
final class AuthMiddleware {
    private let logger: Logger

    init() {
        self.logger = getLogger("AuthMiddleware")
    }

    func authorize(token: String) -> User? {
        do {
            return try validateToken(token)
        } catch {
            logger.warn("unauthorized: \(error)")
            return nil
        }
    }
}

/// Token-bucket rate limiter.
final class RateLimitMiddleware {
    private var counts: [String: Int] = [:]
    private let limit: Int

    init(limit: Int) {
        self.limit = limit
    }

    func allow(key: String) -> Bool {
        let n = (counts[key] ?? 0) + 1
        counts[key] = n
        return n <= limit
    }
}

File: benchmarks/fixtures/webapp_kt/src/main/kotlin/webapp/middleware/AuthMiddleware.kt
package webapp.middleware

import webapp.auth.validateToken
import webapp.models.User
import webapp.util.Logger
import webapp.util.getLogger

/** Rejects requests without a valid token before they reach a handler. */
class AuthMiddleware {
    private val logger: Logger = getLogger("AuthMiddleware")

    fun authorize(token: String): User? {
        return try {
            validateToken(token)
        } catch (e: Exception) {
            logger.warn("unauthorized: $e")
            null
        }
    }
}

/** Token-bucket rate limiter. */
class RateLimitMiddleware(private val limit: Int) {
    private val counts: MutableMap<String, Int> = mutableMapOf()

    fun allow(key: String): Boolean {
        val n = (counts[key] ?: 0) + 1
        counts[key] = n
        return n <= limit
    }
}
